### ISSUE TITLE
Structural pattern matching on polymorphic arguments forgets type variable binding in some cases

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#3940]
**Describe the bug**

When using `match` on a polymorphic argument, on `case` branches that are subtypes of that polymorphic type, pyright forgets that the type variable has been bound.

**To Reproduce**

In the example below, all the type assignments are correct except `case B(xs): ....`. When the `if isinstance` block is uncommented, the type assigned to `xs` in the `case` changes unexpectedly (but the assigned type is still incorrect).

```python
from typing import Generic, List, TypeVar

T = TypeVar('T')

class A(Generic[T]):
    __match_args__ = ('xs',)
    xs: List[T]

class B(A[T]):
    pass

def f(a: A[T], a_: A[int], b: B[T]) -> None:
    match b: # no problem, case branches aren't subtypes of b's type
        case B(xs): print(xs) # xs: List[T@f]

    match a_: # no problem, not polymorphic
        case B(xs): print(xs) # xs: List[int]
        case A(xs): print(xs) # xs: List[int]

    # if isinstance(a, B):          # Uncommenting this causes pyright to assign the type
    #     a.xs # a.xs: List[T@f]    # `xs: List[T@f] | List[Unknown]` in the case B(xs)
                                    # branch below

    match a:
        case B(xs): print(xs) # xs: List[Unknown] <-- incorrect
        case A(xs): print(xs) # xs: List[T@f]
```

**Expected behavior**

The assigned type should be `List[T@f]`.

**Screenshots or Code**

See above.

**VS Code extension or command-line**

I'm using neovim + lsp-config to view assigned types. If I run pyright on the command line with `strict` checking, then a type error is reported: `Type of "xs" is partially unknown`.

```
» pyright --version
pyright 1.1.267
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.271, which I just published. It will also be included in a future release of pylance.
